### ISSUE TITLE
Move NVWave to C/C++

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#5096]
**Reported by camlorn on 2015-05-18 18:51**
#4877, the speech refactor, is going to need some sort of new, higher-level mixer.  This mixer will need to b implemented in C.  The existing NVWave code is sufficient but is currently Python, necessitating a great deal of complexity and copying to be reused in this manner.

I propose moving the NVWave API into C, using a background thread for all audio processing which is communicated with via a threadsafe queue and a command structure.  The background thread allows large delays when closing devices and, as discussed on IRC, the ability to insert blocks of zeros for Windows 8 audio ducking (something I cannot personally test).
I propose three changes to the interface of the API which should be discussed before I begin coding.
First, we should fix the bit depth at 16 bits.  8 is nearly useless and WaveOut does not provide straightforward ways to query for higher bit depths, meaning that most code should already be using 16-bit audio anyway.
Second, we should replace blocking on the last buffer with blocking for a configurable number of milliseconds.
Third, we should add and encourage code to call a finalize method. 
I propose [Powercores](http://github.com/camlorn/powercores.git) for the threadsafe queue.  This is written by me, released under the Unlicense, and currently header only.  It comes with tests and has been used in Libaudioverse for at least 6 months.
Internally, I want to mix in chunks.  If we wish to block around 100 milliseconds, we can split the audio into 100 MS pieces internally, maintain exactly two buffers, fill both, and block until one of them becomes available for new data.  This will allow lowering the buffer size in synths, as the API will refuse to block until it has at least 100 MS of audio (i.e. 64-sample chunks are fine, we'll append them together in C).  This should result in increased callback accuracy in future.  Because we are using a push-based API, blocking at slightly higher MS thresholds is acceptible.
Device enumeration stays as-is in Python.  This is probably about 300-500 lines of C/C++.  I plan to write the patch.
Let's talk about where I'm either being stupid or need to explain further.  I'm sure I've missed something.

**Comment by jcsteh** [comment]
@camlorn: One other thing i just thought of is that there may be (are?) differences in WASAPI and WinMM device naming. There are several synth drivers which use their own audio output code and these seem to use WinMM device names. This is something that needs further investigation as to whether it can be worked around, etc. The "Microsoft Sound Mapper" device at least will need to be special cased.

You should also be aware that implementation of audio ducking is now going ahead in #3830 and will be included in 2016.1 barring catastrophe. That may impact this work also. The audioDucking module is separate, but nvwave does need to call it and I'm not sure how this will get called in your proposed implementation.

**Comment by camlorn** [comment]
can you provide an example of a synth that uses Winmm device naming?  I need to see this in order to properly conceptualize what's going on and comment further.  Wave mapper is special case anyway, as Wasapi actually doesn't natively provide any equivalent; this is currently unimplemented but allowed for in audio_io's public API.  I think that I'm coincidentally even using the same constant for it.  As I mentioned on twitter, I'm now preparing to turn my attention to the auxiliary parts of the Libaudioverse stack; adding the wave mapper special case properly is one of the things on my short-term todo list.  If worst comes to worst, I can turn off Wasapi support.  I think that providing a migration path to it would be a good idea, however; we've currently got two add-ons that benefit from low-latency audio.
As for audio ducking, I need to see code.  I think there's already a branch, and I'm assuming that it looks like what I saw before.  I'd probably just duplicate the API and fold it into the C portions: I can give the thread the ability to properly insert silence.  This was something I was already planning to support anyway, so I'm not sure how much it changes things.  On a side note, I think I can possibly get NVDA playing nicely with audio enhancements as well.

**Comment by jcsteh** [comment]
> can you provide an example of a synth that uses Winmm device naming?

sapi5. Also, all existing configurations use WinMM device names. While forcing users to reconfigure their audio output is not a showstopper, it's also not ideal if we have a better option.

> As for audio ducking, I need to see code.  I think there's already a branch, and I'm assuming that it looks like what I saw before.

That depends how long ago. It's a lot different as of a couple of weeks ago. The branch is t3830. I'd probably just look at audioDucking.py to see the API and then nvwave.py and synthDrivers/sapi5.py to see how it's called.

> I'd probably just duplicate the API and fold it into the C portions: I can give the thread the ability to properly insert silence.

Note that we only need a delay, not actual 0 data. I actually think this may be simpler to keep in the Python code; remember, nvwave.WavePlayer.feed is still allowed to block. Certainly, the audioDucking API needs to be accessible from Python even if you move it to C++ because of stuff that has its own audio output code. (Sidenote: porting sapi5 to nvwave proves to be much more complicated than one would hope.)

**Comment by camlorn** [comment]
It may, and do note the may, be possible to backtranslate from Wasapi device names to WaveOut device names.  I don't like this personally if I'm going to use audio_io, and it's kind of a hack, but I've seen things in the Wasapi API for it.  As I said, though, Wasapi can be turned off and, at this point, I probably will.  It's a one-line comment and not including some files.
I'll look over audio ducking later today and comment further on that situation.

**Comment by camlorn** [comment]
Ducking looks like not a problem, and can probably just stay in Python.
I'm willing to look at porting Sapi5 to NVWave as part of this work, assuming it's possible and we want it badly enough.

**Comment by jcsteh** [comment]
> It may, and do note the may, be possible to backtranslate from Wasapi device names to WaveOut device names.  I don't like this personally if I'm going to use audio_io, and it's kind of a hack,

Granted it's a hack, but we do have a lot of legacy t deal with here unfortunately.

> Wasapi can be turned off and, at this point, I probably will.

The question is whether you lose too much doing that; e.g. it'd suck if we had to keep buffers high to deal with certain crappy hardware.

> Ducking looks like not a problem, and can probably just stay in Python.

Makes sense to me.

> I'm willing to look at porting Sapi5 to NVWave as part of this work, assuming it's possible and we want it badly enough.

It'd be good to have, but Mick spent a good few days on this and kept hitting problems. There are a whole heap of weird COM interfaces and it looks like even when you try to just proxy them, things start breaking. There is some sort of memory stream thing you can use, but I'm guessing we'd have to wait until audio finishes dumping before we can play the chunk, which would be horrible. Anyway, this is probably a discussion to have at some later point and somewhere else. :)

**Comment by camlorn** [comment]
This keeps getting dropped because we keep discussing it offline, whereupon we hit some new question that wasn't thought of.  If there are any users around with troublesome hardware, I'd like to talk to them; I suppose I should put out a call on nvda-devel.  I'm thinking maybe some sort of test programs to see what the good parameters are might be useful.  My goal here is to keep discussion moving, even if only for us to move in the direction of waiting a while.  

Anyhow, here are the options as I see them.  If this is a complete picture then I think we might actually finally have a starting point.  I think this incorporates everything we've discussed.  I don't mention audio ducking because I'm assuming it will stay in Python and that the public interface of NVWave will remain compatible, but possibly with extensions.  Anyhow:

Leave it as it is and close this ticket.  We could still get some of the benefits in the speech refactor.  This would involve adding some sort of optional interface to the synth that basically says "I run at 44.1khz and you can send me special mix this audio chunk in" speech commands.  You can pick whatever sample rate you want here.  Obviously precise enough syncing might be an issue.  I think that we'll end up adding extra functionality for this eventually, but we don't have to do this part of it now.  The complexity here is that the synth would have to do the resampling, and I doubt many actual synths implement such a command.  A synth author could still do it, however, by implementing the synth in C (yes, I know this is ugly. I am simply laying out options).  If we think the speech refactor will never happen, this may be the best option.

Reimplement NVWave using exactly the same algorithms in C.  This would be like the above, but performance would be better and a one-for-one translation of the current stuff isn't hard to do.  The biggest advantage is that we don't gain bugs.  Being in C would give us the ability to do stuff like add in a resampler or APIs for sharing the NVWave instance without having to deal with the performance problems.  It would also no longer be fighting the GIL, which may or may not be an issue.  In addition, we can keep the current buffer sizes while allowing synths to submit in smaller chunks, giving us higher accuracy in terms of indexing.  Concatenating audio in C is not very bad at all in terms of performance.

Reimplement the current NVWave in C, then add a Wasapi fast path.  [There is a way to convert Winmm device indexes into Wasapi device identifiers.](https://msdn.microsoft.com/en-us/library/windows/desktop/dd370819%28v=vs.85%29.aspx).  After this, it's roughly the same as above: we add whatever other APIs we need alongside it for synced sound playback.  At the moment, this is what I prefer: we get the advantage of Waasapi but keep all the backwards compatibility with minimal effort.  It appears that Wasapi actually even provides sample rate and channel order conversion, but this is poorly documented and I only stumbled on it by accident.  Assuming it works as advertised, that means an initial version would be relatively quick.  We could implement NVWave as-is in C and then add this later as well, which is possibly appealing.  Wasapi also gives us other possible benefits like device-offloaded streams which are supposed to give you better battery life.

Finally, implement a global mixer.  This one requires some discussion.  The problem with a global mixer is that it would work well for Wasapi but barely at all for Winmm.  The only reason NVDA has good latency at all right now is because devices are opened when playback starts.  If we have a global mixer, there's going to be at least 100MS of latency on everything we do if it's using Winmm, and I consider it premature to assume that Wasapi isn't going to need similar things.  I have preliminary evidence that it doesn't, but not enough data to be sure.  The problem is that you have to keep enough audio in it to keep it from underrunning, and as we've said before you discovered that the good value for that to work everywhere is 300 MS.  Unfortunately, the advantages are everything else we could ever want.

Why I no longer like this option has to do with me gaining experience and learning to think about audio programming better as Libaudioverse continues.  But there are open questions and favorable answers could significantly change my mind.  To put it simply, we maintain at least 2 buffers of a size large enough to not be troublesome.  One of them is playing and one of them is being loaded with data.  Suppose both are 100MS and the playing buffer just started.  Suppose that filling the available one takes 0 MS.  In this case, it will take 100 MS for the new data to reach the API, plus whatever is between it and the speakers.  So, put another way, this hypothetical example gives 100MS on every utterance.  I can think of ways to maybe make a global mixer work well even with Winmm given that our most common use case is one stream playing at a time, but it's complicated and that assumption may be premature.  This option is why I mentioned possible test programs, as narrowing down the Wasapi parameters would be helpful.

It should be mentioned at this point that Wasapi allows retrieving the device period, the rate at which the device grabs new audio from the in-memory buffer.  This is my biggest open question as related to the mixer.  On Vista and later, it might let us figure out the latency requirements even for Winmm.  If true, then we can lower the latency for everyone but those people who would have poor audio experiences anyway.  Do we know if people were having the need for the incredibly high buffer sizes on XP?  Thinking about it as I write this, I'd put my money on it being Vista and 7 users that needed such high buffer sizes.

MSDN specifically says that you can't go below a device period in terms of latency. Winmm is mostly a thin wrapper over Wasapi. NVDA makes buffer synonymous with latency. Q.E.D. NVDA buffer sizes smaller than the device period are a problem.  Or so goes my theory, but if XP is contributing to this issue then we have to figure out how to deal with it if we want a global mixer.

Or, we could just wait.  We haven't needed it yet, the speech refactor doesn't seem to be happening soon as far as I know, and keeping an issue open costs nothing.  But I do want this to move, if slowly, and that movement is on me.

So, what question have I missed this time?  Alternatively, what's the stupid thing I'm saying here?  Do we finally have an accurate picture of the situation?

Sorry this got a little bit disordered.

**Comment by jcsteh** [comment]
I think this is a fairly accurate (if necessarily complicated) summary of where we are. Some thoughts:

* I agree that it's fine for audio ducking to stay in Python.
* Having the ability for NVDA to play audio regardless of whether the synth supports it is a good thing. Thus, I don't think we should close this ticket, though it's not high priority either.
* IMO, reimplementing nvwave as is in C isn't worthwhile. There are some advantages as you say, but I don't think they're sufficient to justify the effort. Yes, it lowers latency, but is it low enough to synchronise audio with speech for the use cases you and others want?
* Does nvwave with WASAPI and no mixing actually solve the problems you want to solve? Yes, it gives us lower latency in most cases, but again, is it low enough to synchronise audio with speech for the use cases you and others want?
* Regarding the global mixer, yes, there'll be latency on everything we do after the start of an utterance, but unless I'm missing something, this isn't actually a problem for the requested use cases. If you want a sound to be played at a particular point in the speech, that will work just fine: there'll be latency, but there's latency in the speech as well, so they're still both synchronised. The only condition to keep this "responsive" is that the first chunk of audio for an utterance after silence needs to be played immediately. It's fine for subsequent chunks to incur latency. When I say utterance after silence, I mean either the first utterance after there hasn't been any audio for a while or the first utterance after audio was interrupted (e.g. key press).
* I actually think 300 ms was more something we needed for Win2k, which we no longer support. It's also entirely possible that this could be lowered somewhat simply by rewriting nvwave in C as you suggest (but see above as to whether this actually solves the proposed use cases). Once we have an implementation that addresses the proposed use cases, we can guineapig lower latencies on next and then master.

**Comment by camlorn** [comment]
Reimplementing NVWave with Wasapi in C solves enough of it that I can 
get Unspoken working.  Whether that's enough for all use cases, I 
honestly have no idea at all.

We could get audio regardless of if the synth supports it by simply 
opening an NVWave instance in the add-on.  This is what Unspoken does 
currently.

I like the mixer.  I think we can sort of have the mixer.  I think that 
the trick with the mixer is to shut it down and flush the queue when all 
the NVWave instances go idle.  But getting this to work in the general 
case is possibly tricky.

But.  If the mixer only uses Wasapi and if none of the devices have high 
periods, then it's fine for it just to keep running. That's why I don't 
like it.  Whatever hack we end up doing would probably be only for 
WaveOut.  In the case of Wasapi, those users who would have high 
latencies are going to have high latencies anyway (a Wasapi mixer gets 
you a latency of one device period. You can't get lower than one device 
period.  Trust me, I learned it the hard way).

Since the summary is accurate, though, I'm going to look at the NVWave 
API again and see what I can see.  If I can get the mixer hack working, 
then the mixer is possibly the best option.

I wish we could profile NVDA's usage in terms of power.  I really do 
think that being in C would help some of the battery life complaints 
that I hear from time to time.  If this does go through, I'd love to see 
some sort of before and after.

**Comment by jcsteh** [comment]
> Reimplementing NVWave with Wasapi in C solves enough of it that I can 
> get Unspoken working.  Whether that's enough for all use cases, I 
> honestly have no idea at all.

Is that because you don't care if latency is equal to device period?

> But.  If the mixer only uses Wasapi and if none of the devices have high 
> periods, then it's fine for it just to keep running.

* IMO, you still want your initial output to be as near to instant as possible. If you leave it running, you will always have latency equal to device period at minimum, even if NVDA hasn't produced any audio for quite a while or you explicitly interrupted it.
* Leaving the device open also causes issues for virtual machines and remote desktop. It seems VMWare (and now VirtualBox) give you lower audio latency if the device isn't left open, but if you leave it open, the latency increases. For NVDA right now, this means you always have higher latency unless you close the tones player, but closing the tones player makes things much snappier.

> (a Wasapi mixer gets 
> you a latency of one device period. You can't get lower than one device 
> period.  Trust me, I learned it the hard way).

You of course would know more about WASAPI than I do, but from what I've learnt from playing with REAPER, there seem to be two latency options for WASAPI: one for exclusive and one for shared. The value for shared is always higher than the one for exclusive, presumably because shared has to go through the kernel mixer but exclusive doesn't. So, what does device period actually refer to here?

> Since the summary is accurate, though, I'm going to look at the NVWave 
> API again and see what I can see.  If I can get the mixer hack working, 
> then the mixer is possibly the best option.

If the latency with WASAPI really is low enough for most use cases, it certainly would be far simpler if we didn't have to mess around with our own mixer. More complex code = more potential for problems. So, if you're happy with WASAPI latency and just opening multiple streams, I think I'd prefer to do that.

If we do abandon the mixer, can you still do all audio processing in one thread; i.e. multiple streams in one thread?

> I wish we could profile NVDA's usage in terms of power.  I really do 
> think that being in C would help some of the battery life complaints 
> that I hear from time to time.  If this does go through, I'd love to see 
> some sort of before and after.

Yeah, that'd be very useful info to have. There must be a way to do it, but I don't know how.

**Comment by camlorn** [comment]
I'm not set up for quoting.  I should go figure that out.  Also, it's hard to talk about this because I think we either have slightly different understandings or are just using words in slightly different ways.  Admittedly there isn't really a standard vocab I know of.

TLDR: You're selling me on the mixer but just redoing NVWave with Wasapi is enough if we want to require extra work on add-ons, we can probably profile NVDA for a comparison, and insert explanation of how Wasapi works and a link to some code that is along the lines of what we need because it's hard to explain how it works in words.  Finally, we have to discuss Winmm unless we don't care about XP or won't in the near future.

If the synth sends through NVDA and I want to output in sync, I can share the WavePlayer instance of the synth.  This requires three things:

- When we refactor speech, we make the speech command for callbacks give me an WavePlayer instance.

- Synths are required not to push more audio (or at least not significantly more audio) before executing the command.

- The WavePlayer instance resamples audio data internally rather than leaving it to whatever.

At this point, you can sidechain audio data.  Your latency doesn't matter at all.  You add a secondary function that blits buffers onto the outgoing stream just before it's pushed.  For synths that don't use NVWave, this fails some, but I'm skeptical to the precision of the indexing of synths that don't use it anyway.  The downside is that embedding something like Libaudioverse becomes basically impossible: I'd have to strip a bunch of code out and essentially rewrite Unspoken.  The upside is that this gives you the potential for literally sample-accurate playback, even if we set the minimum amount of pushed audio before playing to something ridiculous like 20 seconds.

We can lower the latency of single WavePlayer instances to the same latency we'd get with a mixer though.  The idea of a buffer or chunk doesn't have to exist.  No matter how we do this, latency is going to be in seconds, probably exposed as a double.

Or we do the mixer, we get everything we want, but the code is harder.  Yes, mixers and only mixers give you the one thread for all streams, but we usually only have one stream anyway and there are other ways to fix tones.py.  Of course this means I kind of just obliterated my own argument about not doing a mixer, if we assume that the only concern is code quality; the mixer option simply takes longer initially.  And we can do the above if we want it, even with the mixer architecture.

In a typical audio architecture, the audio device periodically reads an in-memory buffer every so often, pulling the data and updating some counters to say that space is available again.  The device period is the duration (I believe in nanoseconds) between these checks.  It is usually very small.  The actual buffer can be bigger.  What you have to do is keep at least one period ahead of you at all times, because you don't know when the device is going to check again, exactly.  OS scheduling or any of a myriad of things might make you take longer.  So we usually set the latency higher.  If you go too low, you get underruns: the audio device can and will pull audio out from under you, even if you didn't write any.  Which is why the DAW can crackle and stuff.

This is a description of exclusive mode.  In shared mode, the latency is a bit more because you have one global period for the OS's audio mixer, and that's typically higher.  You also have to make your app agree with the OS's format, rather than demanding that the device agree with your app--this means that you have a higher chance of needing to introduce a resampler or channel mixer.  Wasapi also places a few other restrictions on exclusive mode audio apps that don't apply in shared mode, but they're beyond the scope of the discussion here because we absolutely don't want NVDA using exclusive mode (it owns the sound card and prevents other apps from working.  Also, users can and some will disable exclusive mode; this is actually a reasonable thing to do in some cases).

But what this essentially means is that if you get lower latency than one period even with the current architecture, then it's luck.  I highly doubt that Windows makes the audio card poll separately for every app, but can't prove it; your sound card would essentially have to know about apps on Windows, among other things.  This means, essentially, that there's one global period.  It happens every so often, and the soonest your audio can be heard is at the next one, even if you're pushing.  That could be right now, in which case it's instant.  Or the period might have just happened, in which case it's going to take 99% of a period.  We have evidence that I am right on this account: we don't have NVDA users complaining about how NVDA is randomly latent to a noticeable degree, not as far as I know anyway.

Your point about VMWare is a valid one as well.  I'm just offering explanations and do fully understand that the mixer has to be able to shut itself down.  Given the effort for this, I really don't want to go ahead on anything unless at least the two of us are in agreement that whatever I'm going to do is worth it.

[Here's a link to the audio_io Wasapi backend.](https://github.com/camlorn/audio_io/blob/master/src/audio_io/wasapi/output_device.cpp).  You want to look at `WasapiOutputDevice::wasapiMixingThreadFunction`.  If we go down the mixer path, this code would be the starting point.  I can't adopt audio_io as-is for our purposes, but this is probably better than me trying to explain what Wasapi looks like.  The one important thing to note is that the padding will update periodically, not continuously: specifically it will decrease by period-in-samples every period-in-seconds.

We can't drop support for Winmm unless NVDA is or has dropped support for XP.  If this happened while I'm not looking, I'm happy to go ahead with just Wasapi.  Having to also do this for Winmm is indeed a maintenance burden.  But assuming that XP still has to be on the table is why I'm discussing it,  and saying it might be a problem.

We can do one simple thing for profiling.  When the work is done, take the before NVDA and the after NVDA.  Have them both read a fixed passage to the end.  Check the CPU time on each.  This would at least tell us if there was a noticeable difference in CPU usage.

And I'm still somewhat concerned about device index compatibility with Winmm, for the simple reason that just because there's code on MSDN doesn't mean it always works.

**Comment by jcsteh** [comment]
> If the synth sends through NVDA and I want to output in sync, I can share the WavePlayer instance of the synth.  This requires three things:
> 
> - When we refactor speech, we make the speech command for callbacks give me an WavePlayer instance.

If you do things this way, isn't that the same as the mixer except it doesn't transparently merge multiple WavePlayer instances? As I understand it, that's just an API difference, not a low level difference: at the end of the day, you still have to resample and merge buffers.

In my last comment, my main question to you was actually this: is the latency with WASAPI low enough that you don't care even if the two streams were entirely separate? That is, the sound you play would always be a bit late relative to the chunk of speech, but would the latency with WASAPI be low enough that this isn't important for your use cases? Right now, that latency is too high (300 ms), but would, say, 100 ms (maybe lower if we could get it) be too high?

Regarding XP, yes, we do still need to support XP. However, IMO, if sound isn't synchronised well with speech on XP, I don't care. The key point is that we can't regress XP, but if a new feature doesn't work on XP or doesn't work as well, I don't care.

**Comment by camlorn** [comment]
It should be, yes.  I think the cases in which it's not are going to be cases where the audio experience qualifies as horrible anyway.  I have reason to believe that 50MS or so is good enough, but I want more accuracy if I can get it.
The mixer is accurate to one period at minimum.  Sharing the WavePlayer instance that the synth has gives the possibility to be accurate to one sample.  Obviously few synths are actually this accurate with indexing.  The trade-off is that being more accurate makes it harder and harder on add-on developers.  But having a mixer doesn't close this possibility anyway, should we ever decide to do it: the features are pretty much completely orthogonal.
The problem that has to be overcome in any case is that there's no way for an add-on to be more accurate than `period+synth_latency`, where `synth_latency` is presumably set by the synth to something like 50MS or 100MS.  Honestly, though, we can overcome this at a later stage in the work.  I think getting a global mixer is the first step at this point.
If we presume that NVDA will never need these features itself, we can keep all the code we have now as an XP fallback and let the Python module choose which to import.  This means that newer add-ons will outright crash on XP, but I don't personally see that as a problem as I'm the only one who has come forward with an intent to use this stuff.  Since this is something you seem okay with, I think we have a plan.  Presumably NVDA will drop XP support in the next few years (or, well, eventually), and we can just kill the old code then.

**Comment by jcsteh** [comment]
> The mixer is accurate to one period at minimum. Sharing the WavePlayer instance that the synth has gives the possibility to be accurate to one sample.

I feel like we keep going around in circles on this point. As I've tried to explain before, I'm imagining that with the mixer, most WavePlayer instances wouldn't manage audio output themselves. They'd just feed data to the mixer and the mixer would merge the buffers at that point. In terms of what actually happens at the audio level, the mixer should be no different to sharing the WavePlayer instance. The difference is the architecture: with the mixer, most WavePlayers basically become a thin proxy which forwards stuff to the mixer. Can you explain why you see these as so different and why the mixer should incur greater latency than your idea of sharing the WavePlayer instance?

Regarding XP, NVDA may very well need to be able to play sounds in sync with speech in the near-term. Also, it is absolutely not okay for new add-ons to crash on XP. I'm fine with latency being significantly higher. At worst, we can say sounds won't work at all and affected add-ons will need to prevent installation for XP.

**Comment by camlorn** [comment]
When I execute a callback speech command with the mixer, I don't know how close the speech is to being outputted.  It's not that the latency is increased but that the accuracy is decreased.  Depending on how long my callback takes and a myriad of other factors, you can be more or less synchronized and there's no good way to tell.  The maximum error in this case is the period of the mixer plus the latency the mixer introduces plus the latency introduced by the WavePlayer instance plus the error of the synth's indexing.  It is actually theoretically possible to be early as well as late, and with the current numbers this is what would happen in the common case.  I think that the open question here boils down to what the bound on this error actually is in practice, and I think the only way to know is to start doing the code and tweaking things until we see if we need it.
A WavePlayer instance must already integrate a resampler and be able to hold its audio back until it's over the configured threshold, which are both things I thought we had already discussed.  Consequently, much of the infrastructure to do this second super-accurate interface is already in place, which is why I'm still suggesting we consider adding it later.  It also has the advantage of working well even when using WaveOut on XP, and could flat out enable us to just write a mixer on WaveOut if people were okay with using it (which isn't out of the question-it could be made convenient for the common case of playing an audio file alongside speech).
Does this clear it up any?  I suspect you will say no...

**Comment by jcsteh** [comment]
#7599 provides the core of speech refactor, including callbacks. `BeepCommand` and `WaveFileCommand` have been implemented using callbacks. espeak, oneCore and sapi5 have all been updated to provide much more accurate indexing. From a screen reading eprspective, this seems pretty accurate to me.

Regarding indexing for espeak and oneCore, `nvwave.WavePlayer.feed` takes an `onDone` callback which gets called when this chunk finishes playing. So, the audio we've pushed is obviously ahead of the index, but the callback is at the right point in the output. This should also make it much easier for other drivers to get this right.

Note that indexing isn't quite as accurate with espeak + sonic (rate boost) because espeak screws up the position data in this case.

With this approach, based on what you're saying above, we will still encounter minimum device period latency after the callback if we push audio using a separate channel (relying on the Windows mixer). My feeling is that this doesn't matter in practice for screen reading, based on my tests with #7599.

I'd suggest playing with `WaveFileCommand`, etc. in #7599 and seeing what you think from a practical perspective. That will help to figure out what (if anything) needs to be done here.

Note that purely porting nvwave as it is now to C++ might still be good from a CPU/power perspective. We'd want to be able to profile that somehow, though.

**Comment by camlorn** [comment]
I'll see if I can play with it at some point in the near future, but I'm now cursed with a dayjob that's going to be on fire for maybe another month to month and a half and haven't got much spare time for anything at all.

I suspect things will still be needed.  Given that we're dropping XP in the near future, porting NVWave for i.e. power isn't necessarily as hard as it used to be.  If we assume (probably incorrectly) that the ecosystem around Wasapi audio libs has improved, it may just be a drop-in sort of thing.

You may have fixed my other concerns, though there may be a side issue of still needing to disable control announcements for something like Unspoken.

is it possible to get callbacks in *before* controls are spoken?  I'm assuming yes.

I think this will boil down to porting Unspoken to use the new system and seeing how good and/or bad it is.

**Comment by jcsteh** [comment]
The injection of callbacks for unspoken is out of scope for this issue, but
briefly, you need some stuff we don't yet have:
1. getSpeechTextForProperties needs to return a speech sequence; and
2. getSpeechTextForProperties needs to have extension point filters for
both its input and its output.
The groundwork has been done to make this possible, but these two pieces
themselves haven't been done yet. With those two pieces, you have
everything you need to inject callbacks before control information.

As I say, though, that should be addressed in a separate issue.

**Comment by FelipeZanabria** [comment]
How feasible can you use un4seen's Bass library? It has a sub library that supports wasapi, and has examples in c/c++, as well as Python (pybass).
Here is the url:
https://www.un4seen.com/bass.html
